reactor-core/src/main/java/reactor/core/publisher/FluxDistinctUntilChanged.java

```diff
@@ -103,10 +111,26 @@ public boolean tryOnNext(T t) {
 				return true;
 			}
 
-			if (Objects.equals(lastKey, k)) {
+			if (null == lastKey) {
 				lastKey = k;
+				actual.onNext(t);
+				return true;
+			}
+
+			boolean equiv;
+
+			try {
+				equiv = keyComparator.test(lastKey, k);
+			}
+			catch (Throwable e) {
+				onError(Operators.onOperatorError(s, e, t));
+				return true;
+			}
+
+			if (equiv) {
```

Comment:
I'm not sure this should have been merged as is. This `equiv` block means that the oldest equivalent key will always be the one compared.

If your criteria is "the 2 keys are equivalent if within +- 2 from each other" (`abs(abs(lastKey) - abs(key)) > 2`) and you get `--3--4--5--2--|->" sequence, it will compare 3 and 4, 3 and 5 then _3 and 2_ and will only output `--3--|->`.

Whereas by always replacing the `lastkey` in the equiv block with `k`, you'd get the following comparisons:
 3 and 4, 4 and 5, 5 and 2 (changed!) => output `--3--2--|->`.

I'm not sure which of the 2 is the correct (or more desirable) behavior @osi @smaldini?

If the current behavior seem like the best, keep it that way. Otherwise we'll need to open a follow up issue.